# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUserFontSet.h
# Commit: 1b1a7d2f27c4
# Full Hash: 1b1a7d2f27c4428fb8a9ff42f0ba361ec0bb56e6
# Author: Cameron McCormack <cam@mcc.id.au>
# Date: 2019-06-07 09:47:19
# Regressor Bug: 1490792
# File Overlap Count: 1
# Description:
#   Bug 1490792 - Part 4: Perform OpenType sanitization OMT. r=jfkthame
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D33897
# ==============================================================================

diff -r f38b9eb31da7 -r 1b1a7d2f27c4 gfx/thebes/gfxUserFontSet.h
--- a/gfx/thebes/gfxUserFontSet.h	Thu Jun 06 18:23:31 2019 +0000
+++ b/gfx/thebes/gfxUserFontSet.h	Thu Jun 06 19:08:52 2019 +0000
@@ -10,8 +10,10 @@
 #include "gfxFontFamilyList.h"
 #include "gfxFontSrcPrincipal.h"
 #include "gfxFontSrcURI.h"
+#include "nsProxyRelease.h"
 #include "nsRefPtrHashtable.h"
 #include "nsCOMPtr.h"
+#include "nsIFontLoadCompleteCallback.h"
 #include "nsIMemoryReporter.h"
 #include "nsIPrincipal.h"
 #include "nsIRunnable.h"
@@ -22,6 +24,8 @@
 #include "mozilla/ServoStyleConsts.h"
 #include "mozilla/net/ReferrerPolicy.h"
 #include "gfxFontConstants.h"
+#include "mozilla/LazyIdleThread.h"
+#include "mozilla/StaticPtr.h"
 
 class gfxFont;
 
@@ -635,6 +639,8 @@
     MOZ_ASSERT_UNREACHABLE("not meaningful for a userfont placeholder");
   }
 
+  static void Shutdown() { sFontLoadingThread = nullptr; }
+
  protected:
   const uint8_t* SanitizeOpenTypeData(const uint8_t* aData, uint32_t aLength,
                                       uint32_t& aSaneLength,
@@ -650,12 +656,11 @@
 
   // when download has been completed, pass back data here
   // aDownloadStatus == NS_OK ==> download succeeded, error otherwise
-  // returns true if platform font creation sucessful (or local()
-  // reference was next in line)
   // Ownership of aFontData is passed in here; the font set must
   // ensure that it is eventually deleted with free().
-  bool FontDataDownloadComplete(const uint8_t* aFontData, uint32_t aLength,
-                                nsresult aDownloadStatus);
+  void FontDataDownloadComplete(const uint8_t* aFontData, uint32_t aLength,
+                                nsresult aDownloadStatus,
+                                nsIFontLoadCompleteCallback* aCallback);
 
   // helper method for creating a platform font
   // returns true if platform font creation successful
@@ -663,12 +668,32 @@
   // ensure that it is eventually deleted with free().
   bool LoadPlatformFontSync(const uint8_t* aFontData, uint32_t aLength);
 
-  // helper method for LoadPlatformFontSync
+  void LoadPlatformFontAsync(const uint8_t* aFontData, uint32_t aLength,
+                             nsIFontLoadCompleteCallback* aCallback);
+
+  // helper method for LoadPlatformFontAsync; runs on the FontLoader thread
+  void StartPlatformFontLoadOnWorkerThread(
+      const uint8_t* aFontData, uint32_t aLength,
+      nsMainThreadPtrHandle<nsIFontLoadCompleteCallback> aCallback);
+
+  // helper method for LoadPlatformFontAsync; runs on the main thread
+  void ContinuePlatformFontLoadOnMainThread(
+      const uint8_t* aOriginalFontData, uint32_t aOriginalLength,
+      gfxUserFontType aFontType, const uint8_t* aSanitizedFontData,
+      uint32_t aSanitizedLength,
+      nsMainThreadPtrHandle<nsIFontLoadCompleteCallback> aCallback);
+
+  // helper method for LoadPlatformFontSync and
+  // ContinuePlatformFontLoadOnMainThread; runs on the main thread
   bool LoadPlatformFont(const uint8_t* aOriginalFontData,
                         uint32_t aOriginalLength, gfxUserFontType aFontType,
                         const uint8_t* aSanitizedFontData,
                         uint32_t aSanitizedLength);
 
+  // helper method for FontDataDownloadComplete and
+  // ContinuePlatformFontLoadOnMainThread; runs on the main thread
+  void FontLoadFailed(nsIFontLoadCompleteCallback* aCallback);
+
   // store metadata and src details for current src into aFontEntry
   void StoreUserFontData(gfxFontEntry* aFontEntry, bool aPrivate,
                          const nsACString& aOriginalName,
@@ -715,6 +740,8 @@
   gfxUserFontSet* MOZ_NON_OWNING_REF
       mFontSet;  // font-set which owns this userfont entry
   RefPtr<gfxFontSrcPrincipal> mPrincipal;
+
+  static mozilla::StaticRefPtr<mozilla::LazyIdleThread> sFontLoadingThread;
 };
 
 #endif /* GFX_USER_FONT_SET_H */